pkg/util/pods.go

```diff
@@ -332,20 +323,67 @@ func podJobIsFailed(pod *corev1.Pod) bool {
-// Init containers are also checked because they can be declared in template
-// tests, but those added by the test infrastructure are ignored.
-func podHasStarted(pod *corev1.Pod) bool {
-	if pod.Status.Phase == corev1.PodRunning {
-		return true
+// checkPending checks if a pod has been pending for too long
+// The purpose of this function is to cause a failure when a pod is found to be
+// potentially permanently "pending", so that the test fails earlier than its
+// timeout period, which is usually much longer to accommodate legitimate tests.
+//
+// A pod can be in one of the following states (as described in
+// https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle#pod-phase):
+//
+//   - Running/succeeded/failed: all containers have successfully started, no
+//     error is returned.
+//   - Unknown: pod state cannot be recovered, likely indicating a cluster
+//     failure.  This state has been deprecated since Kubernetes 1.22
+//     (https://relnotes.k8s.io/?kinds=deprecation&releaseVersions=1.22.0).
+//     This implementation chooses to ignore it.
+//   - Pending: either "init" containers are being executed or they are done and
+//     not all containers have started.  The verification described below is
+//     performed.
+//
+// Individual containers in a pending pod can be in either the waiting,
+// running, or terminated state.  Only those in the first state are considered
+// by this function.  The verification performed is different for "init" and
+// regular containers:
+//
+//   - "Init" containers execute serially, so they are checked in sequence.  If
+//     any is found to be "waiting" for more than the maximum period, relative
+//     to the finishing time of the previous (or to the creation time of the
+//     pod, when checking the first), an error is returned.
+//   - Regular containers are started in parallel.  If any is found to be
+//     "waiting" for more than the maximum period, relative to the finishing
+//     time of the last "init" container (or to the creation time of the pod,
+//     if there are none), an error is returned.
+func checkPending(pod corev1.Pod, timeout time.Duration, now time.Time) error {
+	if pod.Status.Phase != corev1.PodPending {
+		return nil
 	}
-	// Status is still `Pending` while init containers are executed.
+	check := func(t time.Time, name string) error {
+		if now.Before(t.Add(timeout)) {
+			return nil
+		}
+		return fmt.Errorf("container %q has not started in %s", name, timeout)
+	}
+	prev := pod.CreationTimestamp.Time
 	for _, s := range pod.Status.InitContainerStatuses {
-		if s.Name != "cp-secret-wrapper" && s.State.Running != nil {
-			return true
+		if s.State.Running != nil {
+			return nil
+		} else if w := s.State.Waiting; w != nil {
+			return check(prev, s.Name)
+		} else if t := s.State.Terminated; t != nil {
+			prev = t.FinishedAt.Time
+		} else {
+			panic(fmt.Sprintf("invalid container status: %#v", s))
```

Comment:
Do we ever `recover()` somewhere or does this cause the entire application to fail?

Reply:
The Kubernetes API guarantees that exactly one of the three states will be present.  This is an `assert`ion to detect unsatisfied preconditions, either from a real cluster or in tests.

Reply:
`panic()`-ing [disrupts the regular execution flow](https://go.dev/blog/defer-panic-and-recover) in favor of a `try {} catch {}`-ish style. It has definitely its use cases but I would rather prefer to return and handle a regular `error`, unless you are planning to start converting the codebase into something that follows [this](https://en.wikipedia.org/wiki/Design_by_contract).

Reply:
There are two possibilities for this code to be reached:

- Synthetic data which are incorrect.
- Data from a server which has violated its promise, such that there is no telling what the actual state is.

I don't see how these can be recovered from and what the calling code would do to continue operating.

Reply:
> possibilities for this code to be reached:
- kubernetes introduces a new state for the containers.

As you said, it's likely that such an error can't be recovered but to me, giving the caller the choice to decide what to do with an `err` is more flexible than being abruptly terminated by the callee. By the way, such a "low level" detail is not worth blocking an entire PR, so...